### ISSUE TITLE
iOS Support planned?

### ISSUE THREAD

**Issue by eduo** [bvschaik/julius#643]
Ask your question here.
I was wondering if iOS or iPadOS is in the plans at all. With Mac compilation done, it may be closer to possible and with the recently improved support for keyboard and mouse it's not even necessary to support touch 100%.

**Comment by crudelios** [comment]
There are no plans to support iOS.

Supporting iOS not only requires a developer license which costs about 100 dollars a year, but the Apple Store's terms of use are not compatible with Julius' license (GPL v3).

**Comment by eduo** [comment]
> There are no plans to support iOS.
> 
> Supporting iOS not only requires a developer license which costs about 100 dollars a year, but the Apple Store's terms of use are not compatible with Julius' license (GPL v3).

Just to clarify: Supporting iOS doesn't require a paid license nor does it require violating the GPL. I think you're referring to publishing in the App Store, which I agree with.

I skipped it because I thought it was obvious, my mistake. I meant for sideloading in the same way as spiritually similar projects like Exult or DevilutionX (or ScummVM or RetroArch or ppsspp or many others) currently support compilation for iOS/iPad and in some cases even produce a ready-made IPA that can be sideloaded without ever paying a developer license or publishing in the app store.

I'm sorry I wasn't clear enough. I understand why projects like these wouldn't want to get anywhere near Apple's App Store and I fully support that decision. I meant something else.

**Comment by eduo** [comment]
I appreciate the time taken to give a response. I asked precisely because I saw Android support and guessed most of the UX part could be translated (I understand this is but a small part of the work that would have to be done).

I also asked because I have friends interested that might have the chops necessary to help with such a port. I'll see about them creating a fork and if successful will tell them to make a PR.

Thanks for your efforts and work.

**Comment by bvschaik** [comment]
Before any work is done, I'd like to have a few points clarified. Keep in mind: I did a bit of iOS programming more than 10 years ago (iOS 4/5), so my knowledge is limited and quite outdated...

1. How does this sideloading stuff / install apps without the app store work? How should users install the app and how should we provide it?
2. What about the assets from the original game? Android provides a real filesystem outside the app data, which the user can access and copy things onto from a computer. Within the app you then grant permission to this directory so the game can use it. I don't remember iOS having a similar concept: all data is stored in the app itself: how does the user get the assets there, and the savegames/maps out?
3. Resolution for iPhone will probably be a usability issue: Julius requires at least 640x480 resolution. The largest (effective) iPhone resolution is [736x414](https://developer.apple.com/library/archive/documentation/DeviceInformation/Reference/iOSDeviceCompatibility/Displays/Displays.html), which means we have to downscale. That happens to be about the same scale ratio as the Android phone I have, and the game is already barely playable on that 6" phone (hard to touch buttons).

**Comment by eduo** [comment]
> Before any work is done, I'd like to have a few points clarified. Keep in mind: I did a bit of iOS programming more than 10 years ago (iOS 4/5), so my knowledge is limited and quite outdated...

I'll try to provide some answers.

> 1. How does this sideloading stuff / install apps without the app store work? How should users install the app and how should we provide it?

For Jailed devices three main methods exist (without involving third parties, since some services out there will serve as secondary app stores themselves): Building the app yourself and sideloading a pre-compiled IPA (like an APK but for iPhones). Provenance goes into some detail for these [here](https://wiki.provenance-emu.com/installation-and-usage/installing-provenance).

For sideloading the most common methods nowadays are [Sideloadly](https://sideloadly.io/) and [AltStore](https://altstore.io/). Sideloadly is a single app where you load the downloaded IPA and sign it and send it to the device. Altstore installs a "signing server" in a PC or Mac and wirelessly signs the IPAs you download and refreshes them periodically.

In all cases a free developer account from Apple is required for signing the IPAs. This is how most people sideload nowadays. Free developer accounts require periodical resigning but altstore does this automatically over WiFi.

> 2. What about the assets from the original game? Android provides a real filesystem outside the app data, which the user can access and copy things onto from a computer. Within the app you then grant permission to this directory so the game can use it. I don't remember iOS having a similar concept: all data is stored in the app itself: how does the user get the assets there, and the savegames/maps out?

This is a common thing with engine remakes. For example when DevilutionX was ported to iOS those files were initially only copied via [itunes/finder  file copy](https://www.apeaksoft.com/itunes/itunes-file-sharing.html) 

In [this thread](https://github.com/diasurgical/devilutionX/pull/137) it was discussed adding support for native file handling within the app and the files.app, so the file downloaded from somewhere or copied from a server could be moved into the devilutionX folder. 

This is also how Exult works, and that [expects a whole folder structure](https://www.reddit.com/r/Ultima/comments/mp03s5/ultima_7_exultipad_for_dummies/gu6wt3n/). Essentially it's how most emulators work, too. I posted on reddit some months ago [some of my favorite ones](https://www.reddit.com/r/AltStore/comments/p4ih1f/my_preferred_emulators_all_loaded_through/), all of which work jailed by allowing an IPA to be downloaded, and all of which require files to be put in specific locations.

You're right in that apps can't access external directories, but the app's directory is accessible by the file manager (files app) so files can be put in and out by it.

> 3. Resolution for iPhone will probably be a usability issue: Julius requires at least 640x480 resolution. The largest (effective) iPhone resolution is [736x414](https://developer.apple.com/library/archive/documentation/DeviceInformation/Reference/iOSDeviceCompatibility/Displays/Displays.html), which means we have to downscale. That happens to be about the same scale ratio as the Android phone I have, and the game is already barely playable on that 6" phone (hard to touch buttons).

I believe this is really an ergonomy issue in any smartphone, to be honest. Not so much due to scaling and resolution but due to these things being designed for tiny mouse cursors in huge (by comparison) screens. In these cases the usual is that while the app/game/emulator is available to iPhones, they might be optimized for iPads. Particularly if they were designed for mouse input. For example MiniVMac which emulates a Mac or iAmiga can run on iPhones, but unless you're using an external display, keyboard and mouse, they're hell on eyes and fingers.

**Comment by bvschaik** [comment]
Closing due to inactivity - if there is any concrete progress, comment and I'll reopen.

**Comment by axmo** [comment]
I have Julius (and Augustus) running on iOS, but my project probably isn't a good candidate for a PR. I don't have much experience with CMake, so instead I had CMake generate an Xcode project, and modified that to build for iOS, (which involves an xcworkspace that also builds SDL and SDL-mixer, because I could not find an iOS release of those that would work on iOS simulator and on device)

If anyone is interested, I can move my iOS-file-picker dialog stuff from the external project over to to platform/iOS, and put up a PR with those changes, but it won't build without the xcworkspace and xcproject changes, and I have no idea how to get those changes into CMake...

**Comment by crudelios** [comment]
Hey @axmo,

I'm not sure about Julius (that's @bvschaik's call), but I wouldn't mind an ios project for augustus, even if it couldn't be added to the app store.

I'm also not sure there's a way to do it any differently than you did. Here's a page that I've had catching dust on an old tab on my phone, maybe it's helpful:

https://discourse.libsdl.org/t/state-of-cmake-ios-for-sdl2/34509/2

**Comment by bvschaik** [comment]
> I'm not sure about Julius (that's @bvschaik's call), but I wouldn't mind an ios project for augustus, even if it couldn't be added to the app store.

I'm with you on this one. I don't mind having Julius build for iOS, just the App Store is out of bounds due to licensing.

So please go ahead!

**Comment by rimas-kudelis** [comment]
As you probably know, Apple was forced to allow third-party app stores by EU legislation. Distributing the game freely might be a lot easier now, at least in EU.

A couple well-known such alternative stores are [AltStore PAL](https://altstore.io/) and [Aptoide](https://ios.store.aptoide.com/). Both are free for the users to install and use, and AltStore PAL has a certain amount of focus on classic games and their emulation, so a reimplementation of Caesar III would fit its theme very well. Plus, their app provides ways to add your own package sources, so the developer does not have to pay either (not sure how that works with Aptoide).

Don't know about iPhones, but having Julius on an iPad would be quite fun, I suppose. :)

**Comment by axmo** [comment]
Update: my plan is to add an option for the CMake Xcode generator to output a buildable iOS project, so there wouldn't need to be any extra manual steps when julius or augustus add files to the project. I'll look into it again over the winter holidays.

**Comment by axmo** [comment]
@crudelios I also have the changes for Augustus, but, as I've already forked Julius, Github will not let me fork Augustus... Let me know your preferred method for getting those to you.

**Comment by crudelios** [comment]
Great job! 👍 Excited for this one.

You can add augustus as a new upstream and then simply create a new branch that points to your augustus upstream, that's how I manage to work with both julius and augustus at the same time!

**Comment by axmo** [comment]
To avoid conflicts / unrelated tree issues in git, I guess I should redo the iOS stuff for Augustus as a merge commit from the Julius upstream instead of what I've actually done, which was implement it all twice as two separate projects... (Really, I did this all backwards and got Augustus working first...)

**Comment by crudelios** [comment]
Yeah that would be a good idea. Tell you what, let's merge this here, then I'll merge the latest Julius changes to augustus, and then you can fix what's left. That way there won't be any conflicts when I later merge julius to augustus.

**Comment by bvschaik** [comment]
> Does Julius build and work on iOS now? If so, maybe it should be added to the supported platform list and made available for download?

It builds and works, BUT: it requires the user to compile and install it on their iPhone/iPad themselves, due to all the restrictions that Apple puts on it. Providing a downloadable ipa won't work since that's not signed.

**Comment by rimas-kudelis** [comment]
Hm, I wonder how other open-source applications solve the problem of iOS distribution. Maybe something like a minimal Patreon support tier would help?

**Comment by bvschaik** [comment]
They don't. I looked at OpenTTD and OpenRTC2 as examples and they both don't have iOS support.

There are actually multiple reasons to not deploy on the App Store:

- Signing, requiring an Apple developer account and lots of time to get everything set up and hope it doesn't break. Note that the Julius project does not and will never take donations.
- The licence of Julius is not compatible with the App Store.
- The need to get the game's assets loaded onto the device. For Android this already is a problem: lots of users not understanding it and giving one star reviews because "the game does not work, it's a scam".
- Apple most likely rejecting the app because they cannot review it properly because of the assets.

Using an alternative app store might make some points no longer valid, but enough problems remain to make it not worth the effort.

**Comment by axmo** [comment]
Yeah, still need to sign the .ipa and submit to Apple for notarization if you want to distribute on AltStore PAL. I suppose it might be possible to provide build artifacts in the releases for people to sign themselves and install via old-style AltStore / Sideloadly? I've never used either of those, so I don't know what's actually involved there.

**Comment by rimas-kudelis** [comment]
Hmm, precompiled but not signed builds sounds like an interesting idea, maybe that would be a good compromise? This way @bvschaik wouldn't need to invest her own money and effort into getting the app signed/notarized for the walled garden that iOS/iPadOS is, but users could still be saved from the trouble of building the project and would only have to sign and sideload it themselves.

It seems that this is the way at least some OSS projects are being distributed, such as [DolphiniOs](https://dolphinios.oatmealdome.me), the developer of which also blogged about why they can't distribute via App Store or its EU alternatives.

**Comment by eduo** [comment]
This is indeed the standard process by which most apps are sideloaded nowadays. An unsigned IPA sideloaded by one of various methods.

I wrote [this post in Reddit](https://www.reddit.com/r/AltStore/comments/p4ih1f/my_preferred_emulators_all_loaded_through/) four years ago which essentially lists all the IPAs that I used to use (there're are a couple of source code to compile, but it's mostly IPAs).

AltStore (the sideload version, not the PAL version) even allowed repositories to automatically download and update these IPAs, but the simplest was always to just download and install on top of the existing one.

Most of the emulators now in the store, still offer an IPA to download or did until they didn't have to.
